File: week6/14503/14503.cpp
#include <iostream>
#include <vector>

using namespace std;

int N, M;
int r, c;
int dir;
vector<vector<int> > board;

void set_input();
int clean_all();
void print_board();


int main() {
    ios_base::sync_with_stdio(0);
    set_input();
    int total_clean = clean_all();
    cout << total_clean << endl;
}

void set_input() {
    cin >> N >> M;
    cin >> r >> c >> dir;
    board.reserve(N);
    for (int i=0; i<N; i++) {
        board[i].reserve(M);
        for (int j=0; j<M; j++) {
            cin >> board[i][j];
        }
    }
}

int clean_all() {
    vector<vector<int> > visit(N, vector<int>(M, 0));
    int dr[4] = {-1, 0, 1, 0};
    int dc[4] = {0, 1, 0, -1};

    visit[r][c] = 1;
    int total_clean = 1;

    int nr, nc;
    while (true) {
        bool can_clean = false;
        // check space
        for (int i=0; i<4; i++) {
            dir = (dir + 3) % 4;
            nr = r + dr[dir];
            nc = c + dc[dir];
            if (board[nr][nc] != 1 && visit[nr][nc] == 0) {
                can_clean = true;
                visit[nr][nc] = 1;
                total_clean++;
                r = nr;
                c = nc;
                break;
            }
        }
        // no space to clean
        if (!can_clean) {
            r += dr[(dir + 2) % 4];
            c += dc[(dir + 2) % 4];
            if (board[r][c] == 1)
                break;
        }
    }
    return total_clean;
}

void print_board() {
    for (int i=0; i<N; i++) {
        for (int j=0; j<M; j++) {
            cout << board[i][j] << " ";
        } cout << endl;
    } cout << endl;
}

File: week6/14503/14503.py
import sys
input = sys.stdin.readline

class P_14503:
    def __init__(self):
        self.N, self.M = map(int, input().split())
        self.r, self.c, self.d = map(int, input().split())
        self.board = []
        for _ in range(self.N):
            self.board.append(list(map(int, input().split())))

    def run(self):
        """
        self.d
        - 0: 북
        - 1: 동
        - 2: 남
        - 3: 서
        """
        dr = [-1, 0, 1, 0]
        dc = [0, 1, 0, -1]
        visit = [[0 for _ in range(self.M)] for _ in range(self.N)]

        visit[self.r][self.c] = 1
        total_clean = 1
        while True:
            # check space
            can_clean = False
            for _ in range(4):
                self.d = (self.d + 3) % 4
                nr = self.r + dr[self.d]
                nc = self.c + dc[self.d]
                if self.board[nr][nc] != 1 and visit[nr][nc] == 0:
                    can_clean = True
                    visit[nr][nc] = 1
                    total_clean += 1
                    self.r = nr
                    self.c = nc
                    break

            # if no space to clean
            if not can_clean:
                self.r += dr[(self.d + 2) % 4]
                self.c += dc[(self.d + 2) % 4]
                if self.board[self.r][self.c] == 1: # finish
                    break
        print(total_clean)


if __name__ == "__main__":
    P = P_14503()
    P.run()